### PR TITLE
Add `dhall lint` support for deprecating `Optional/{fold,build}`

### PR DESCRIPTION
**Pull request by Gabriel439**

### CODE REVIEW

**Comment by Gabriel439** [comment]
@sjakobi: I'll postpone this until after the 1.29 release, because addressing your feedback is tricky

**Comment by sjakobi** [comment]
> @sjakobi: I'll postpone this until after the 1.29 release, because addressing your feedback is tricky

Feel free to merge as-is then. We can still add the other refactoring at a later stage.

File: dhall/src/Dhall/Lint.hs [hunk 1]
```diff
@@ -42,15 +47,20 @@ import qualified Lens.Family
...
-        <|> fixParentPath        e
-        <|> removeLetInLet       e
+    lowerPriorityRewrite e =
+            fixAssert                e
+        <|> removeUnusedBindings     e
+        <|> fixParentPath            e
+        <|> removeLetInLet           e
+        <|> replaceOptionalBuildFold e
+
+    higherPriorityRewrite = replaceSaturatedOptionalFold
```
Comment (sjakobi) [inline dhall/src/Dhall/Lint.hs]: Ah, now I understand why my request was tricky to implement! :) Cheers!

Comment (Gabriel439) [reply dhall/src/Dhall/Lint.hs]: Yeah, the rewrite is depth-first, so the `Optional/fold` gets replaced before the rewrite has a chance to match on the larger saturated `Optional/fold` expression

**Pull request merged**